<?xml version="1.0" encoding="utf-8"?>
<launch>
	<!-- /////////////////////////////////////// -->
	<!-- Camera rectification inputs and outputs -->
	<!-- \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\ -->

	<!-- ================== -->
	<!-- Color input topics -->
	<!-- ================== -->
	<arg name="rgb_camera_info"               default="/camera/color/camera_info"/>
	<arg name="rgb_image_raw"                 default="/camera/color/image_raw"/>

	<!-- ================== -->
	<!-- Color output topic -->
	<!-- ================== -->
	<arg name="rgb_image_rect"                default="/camera/color/image_rect"/>

	<!-- ================== -->
	<!-- Depth input topics -->
	<!-- ================== -->
	<arg name="depth_camera_info"             default="/camera/depth/camera_info"/>
	<arg name="depth_image_raw"   	          default="/camera/depth/image_raw"/>

	<!-- ================== -->
	<!-- Depth output topic -->
	<!-- ================== -->
	<arg name="depth_image_rect"   	          default="/camera/depth/image_rect"/>


	<!-- /////////////////////////////// -->
	<!-- Registration inputs and outputs -->
	<!-- \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\ -->

	<!-- ============================== -->
	<!-- depth registered output topics -->
	<!-- ============================== -->
	<arg name="depth_registered_camera_info"  default="/camera/registered/camera_info"/> 
	<arg name="depth_registered_image_rect"   default="/camera/registered/reg_rect"/> 
	<arg name="depth_registered_points"       default="/camera/registered/points"/> 

	<!-- /////////////////////////////// -->
	<!-- 			 Nodes 				 -->
	<!-- \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\ -->

	<!-- Nodelet manager -->
	<node pkg="nodelet" type="nodelet" name="image_proc_nodelet_manager" args="manager" />

	<!-- ########################### -->
	<!-- Color rectification nodelet -->
	<!-- ########################### -->
	<!-- Takes an unrectified image stream and its associated calibration parameters, and produces rectified images. -->
	<node pkg="nodelet" type="nodelet" name="color_image_proc_rectify" args="manager" />
		<node pkg="nodelet" type="nodelet" name="color_image_rectification" args="load image_proc/rectify image_proc_nodelet_manager">
		<!-- Input topics -->
        <remap from="camera_info"                        to="$(arg rgb_camera_info)"/>
		<remap from="image_mono"                         to="$(arg rgb_image_raw)"/>

		<!-- Output topic-->
		<remap from="image_rect"                         to="$(arg rgb_image_rect)"/>
	</node>

	<!-- ########################### -->
	<!-- Depth rectification nodelet -->
	<!-- ########################### -->
	<!-- Takes an unrectified image stream and its associated calibration parameters, and produces rectified images. -->
	<node pkg="nodelet" type="nodelet" name="depth_image_proc_rectify" args="manager" />
		<node pkg="nodelet" type="nodelet" name="depth_image_rectification" args="load image_proc/rectify image_proc_nodelet_manager">
		<!-- Input Topics -->
        <remap from="camera_info"                        to="$(arg depth_camera_info)"/>
		<remap from="image_mono"                         to="$(arg depth_image_raw)"/>

		<!-- Output topic -->
		<remap from="image_rect"                         to="$(arg depth_image_rect)"/>
	</node>

	<!-- ########################## -->
	<!-- Depth registration nodelet -->
	<!-- ########################## -->
	<!-- "register" depth image to rgb camera frame. Reprojecting the depths requires the calibration parameters of both cameras and, from tf, the extrinsic transform between them. -->
	<node pkg="nodelet" type="nodelet" name="image_registration" args="load depth_image_proc/register image_proc_nodelet_manager">
		<!-- Input topics -->
        <remap from="rgb/camera_info"                    to="$(arg rgb_camera_info)"/>
		<remap from="depth/camera_info"                  to="$(arg depth_camera_info)"/>
		<remap from="depth/image_rect"                   to="$(arg depth_image_rect)"/>

		<!-- Params for depth registration -->
		<param name="queue_size"						value="20"/>

		<!-- Output topics -->
		<remap from="depth_registered/camera_info"       to="$(arg depth_registered_camera_info)"/> <!-- Should be similar to color/camera_info -->
		<remap from="depth_registered/image_rect"        to="$(arg depth_registered_image_rect)"/> 
	</node>

	<!-- ########################## -->
	<!-- Depth to pointcloud xyzrgb -->
	<!-- ########################## -->
	<!-- combine registered depth image and RGB image into XYZRGB point cloud. -->
	<node pkg="nodelet" type="nodelet" name="image_proc_pointcloud_xyzrgb" args="load depth_image_proc/point_cloud_xyzrgb image_proc_nodelet_manager">
		<!-- Input topics -->
		<remap from="rgb/camera_info"                    to="$(arg rgb_camera_info)"/>
		<remap from="rgb/image_rect_color"               to="$(arg rgb_image_rect)"/>
		<remap from="depth_registered/image_rect"        to="$(arg depth_registered_image_rect)"/>

		<!-- Params for point cloud conversion -->
		<param name="queue_size"						value="20"/>
		
		<!-- Output topic -->
		<remap from="depth_registered/points"            to="$(arg depth_registered_points)"/>
	</node>

</launch>
